# Исследование надежности заемщиков


Во второй части проекта вы выполните шаги 3 и 4. Их вручную проверит ревьюер.
Чтобы вам не пришлось писать код заново для шагов 1 и 2, мы добавили авторские решения в ячейки с кодом. 



## Откройте таблицу и изучите общую информацию о данных

**Задание 1. Импортируйте библиотеку pandas. Считайте данные из csv-файла в датафрейм и сохраните в переменную `data`. Путь к файлу:**

`/datasets/data.csv`

In [1]:
import pandas as pd

data = pd.read_csv(r"C:\Users\Delta-Copm\Documents\Аналитик данных\Проекты\data_2_borrower_reliability_research.csv")

**Задание 2. Выведите первые 20 строчек датафрейма `data` на экран.**

In [2]:
data.head(20)

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   
15         1    -972.364419         26              среднее             1   
16         0   -1719.934226         35              среднее             1   
17         0   -2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0  -10038.818549         48              СРЕДНЕЕ             1   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   
15   женат / замужем                 0      F   сотрудник     0   
16   женат / замужем                 0      F   сотрудник     0   
17  гражданский брак                 1      M   сотрудник     0   
18    вдовец / вдова                 2      F   пенсионер     0   
19         в разводе                 3      F   сотрудник     0   

     total_income                                 purpose  
0   253875.639453                           покупка жилья  
1   112080.014102                 приобретение автомобиля  
2   145885.952297                           покупка жилья  
3   267628.550329              дополнительное образование  
4   158616.077870                         сыграть свадьбу  
5   255763.565419                           покупка жилья  
6   240525.971920                       операции с жильем  
7   135823.934197                             образование  
8    95856.832424                   на проведение свадьбы  
9   144425.938277                 покупка жилья для семьи  
10  113943.491460                    покупка недвижимости  
11   77069.234271       покупка коммерческой недвижимости  
12            NaN                         сыграть свадьбу  
13  130458.228857                 приобретение автомобиля  
14  165127.911772              покупка жилой недвижимости  
15  116820.904450 

**Задание 3. Выведите основную информацию о датафрейме с помощью метода `info()`.**

In [3]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


## Предобработка данных

### Удаление пропусков

**Задание 4. Выведите количество пропущенных значений для каждого столбца. Используйте комбинацию двух методов.**

In [4]:
data.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

**Задание 5. В двух столбцах есть пропущенные значения. Один из них — `days_employed`. Пропуски в этом столбце вы обработаете на следующем этапе. Другой столбец с пропущенными значениями — `total_income` — хранит данные о доходах. На сумму дохода сильнее всего влияет тип занятости, поэтому заполнить пропуски в этом столбце нужно медианным значением по каждому типу из столбца `income_type`. Например, у человека с типом занятости `сотрудник` пропуск в столбце `total_income` должен быть заполнен медианным доходом среди всех записей с тем же типом.**

In [5]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['total_income'].isna()), 'total_income'] = \
    data.loc[(data['income_type'] == t), 'total_income'].median()

### Обработка аномальных значений

**Задание 6. В данных могут встречаться артефакты (аномалии) — значения, которые не отражают действительность и появились по какой-то ошибке. таким артефактом будет отрицательное количество дней трудового стажа в столбце `days_employed`. Для реальных данных это нормально. Обработайте значения в этом столбце: замените все отрицательные значения положительными с помощью метода `abs()`.**

In [6]:
data['days_employed'] = data['days_employed'].abs()

**Задание 7. Для каждого типа занятости выведите медианное значение трудового стажа `days_employed` в днях.**

In [7]:
data.groupby('income_type')['days_employed'].agg('median')

income_type
безработный        366413.652744
в декрете            3296.759962
госслужащий          2689.368353
компаньон            1547.382223
пенсионер          365213.306266
предприниматель       520.848083
сотрудник            1574.202821
студент               578.751554
Name: days_employed, dtype: float64

У двух типов (безработные и пенсионеры) получатся аномально большие значения. Исправить такие значения сложно, поэтому оставьте их как есть.

**Задание 8. Выведите перечень уникальных значений столбца `children`.**

In [8]:
data['children'].unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5], dtype=int64)

**Задание 9. В столбце `children` есть два аномальных значения. Удалите строки, в которых встречаются такие аномальные значения из датафрейма `data`.**

In [9]:
data = data[(data['children'] != -1) & (data['children'] != 20)]

**Задание 10. Ещё раз выведите перечень уникальных значений столбца `children`, чтобы убедиться, что артефакты удалены.**

In [10]:
data['children'].unique()

array([1, 0, 3, 2, 4, 5], dtype=int64)

### Удаление пропусков (продолжение)

**Задание 11. Заполните пропуски в столбце `days_employed` медианными значениями по каждого типа занятости `income_type`.**

In [11]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['days_employed'].isna()), 'days_employed'] = \
    data.loc[(data['income_type'] == t), 'days_employed'].median()

**Задание 12. Убедитесь, что все пропуски заполнены. Проверьте себя и ещё раз выведите количество пропущенных значений для каждого столбца с помощью двух методов.**

In [12]:
data.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

### Изменение типов данных

**Задание 13. Замените вещественный тип данных в столбце `total_income` на целочисленный с помощью метода `astype()`.**

In [13]:
data['total_income'] = data['total_income'].astype(int)

### Обработка дубликатов

**Задание 14. Обработайте неявные дубликаты в столбце `education`. В этом столбце есть одни и те же значения, но записанные по-разному: с использованием заглавных и строчных букв. Приведите их к нижнему регистру.**

In [14]:
data['education'] = data['education'].str.lower()

**Задание 15. Выведите на экран количество строк-дубликатов в данных. Если такие строки присутствуют, удалите их.**

In [15]:
data.duplicated().sum()

71

In [16]:
data = data.drop_duplicates()

### Категоризация данных

**Задание 16. На основании диапазонов, указанных ниже, создайте в датафрейме `data` столбец `total_income_category` с категориями:**

- 0–30000 — `'E'`;
- 30001–50000 — `'D'`;
- 50001–200000 — `'C'`;
- 200001–1000000 — `'B'`;
- 1000001 и выше — `'A'`.


**Например, кредитополучателю с доходом 25000 нужно назначить категорию `'E'`, а клиенту, получающему 235000, — `'B'`. Используйте собственную функцию с именем `categorize_income()` и метод `apply()`.**

In [17]:
def categorize_income(row):
    try:
        if row <= 30000:
            return 'E'
        elif 30001 <= row <= 50000:
            return 'D'
        elif 50001 <= row <= 200000:
            return 'C'
        elif 200001 <= row <= 1000000:
            return 'B'
        else:
            return 'A'
    except TypeError:
        return 'Unknown'


In [18]:
data['total_income_category'] = data['total_income'].apply(categorize_income)

**Задание 17. Выведите на экран перечень уникальных целей взятия кредита из столбца `purpose`.**

In [19]:
data['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

**Задание 18. Создайте функцию, которая на основании данных из столбца `purpose` сформирует новый столбец `purpose_category`, в который войдут следующие категории:**

- `'операции с автомобилем'`,
- `'операции с недвижимостью'`,
- `'проведение свадьбы'`,
- `'получение образования'`.

**Например, если в столбце `purpose` находится подстрока `'на покупку автомобиля'`, то в столбце `purpose_category` должна появиться строка `'операции с автомобилем'`.**

**Используйте собственную функцию с именем `categorize_purpose()` и метод `apply()`. Изучите данные в столбце `purpose` и определите, какие подстроки помогут вам правильно определить категорию.**

In [20]:
def categorize_purpose(row):
    try:
        if 'автом' in row:
            return 'операции с автомобилем'
        elif 'жил' in row or 'недвиж' in row:
            return 'операции с недвижимостью'
        elif 'свад' in row:
            return 'проведение свадьбы'
        elif 'образов' in row:
            return 'получение образования'
    except:
        return 'нет категории'

In [21]:
data['purpose_category'] = data['purpose'].apply(categorize_purpose)

### Шаг 3. Исследуйте данные и ответьте на вопросы

#### 3.1 Есть ли зависимость между количеством детей и возвратом кредита в срок?

In [22]:
data_children = data.pivot_table(index=['children'], columns='debt', values='gender', aggfunc='count', fill_value=0 )

In [23]:
data_children

debt          0     1
children             
0         13028  1063
1          4364   444
2          1858   194
3           303    27
4            37     4
5             9     0

In [24]:
data_children['total'] = data_children[0] + data_children[1]

In [25]:
data_children['percentage_of_non_return'] = data_children[1]/data_children['total']*100


In [26]:
data_children['percentage_of_non_return'] = data_children['percentage_of_non_return'].apply(lambda x: f"{x:.2f}%")

In [27]:
data_children['borrowers_share'] = (data_children['total']/ data_children['total'].sum())*100

In [28]:
data_children['borrowers_share'] = data_children['borrowers_share'].apply(lambda x: f"{x:.2f}%")

In [29]:
data_children.sort_values('percentage_of_non_return')

debt          0     1  total percentage_of_non_return borrowers_share
children                                                             
5             9     0      9                    0.00%           0.04%
0         13028  1063  14091                    7.54%          66.06%
3           303    27    330                    8.18%           1.55%
1          4364   444   4808                    9.23%          22.54%
2          1858   194   2052                    9.45%           9.62%
4            37     4     41                    9.76%           0.19%

**Вывод:**
Клиенты без детей представляют наименьший риск невозврата кредитов, несмотря на то, что составляют самую большую группу заемщиков

Клиенты с одним и двумя детьми имеют значительно более высокий процент невозврата, что указывает на повышение финансового риска при увеличении количества детей.

Клиенты с тремя и четырьмя детьми также показывают высокий процент невозврата, но их доля в общей массе заемщиков меньше, чем у клиентов с одним или двумя детьми.

Клиенты с пятью детьми показывают идеальный результат с 0% невозврата, но их чрезвычайно малая численность не позволяет делать обоснованные выводы.

Количество детей оказывает значительное влияние на вероятность возврата кредита. Чем больше детей в семье, тем выше процент невозврата. Это может быть связано с увеличением финансовых обязательств и расходов в семьях с большим количеством детей.


#### 3.2 Есть ли зависимость между семейным положением и возвратом кредита в срок?

In [30]:
data_family = data.pivot_table(index=['family_status'], columns='debt', values='gender', aggfunc='count' )

In [31]:
data_family

debt                       0    1
family_status                    
Не женат / не замужем   2523  273
в разводе               1105   84
вдовец / вдова           888   63
гражданский брак        3749  385
женат / замужем        11334  927

In [32]:
data_family['total'] = data_family[0] + data_family[1]

In [33]:
data_family['percentage_of_non_return'] = data_family[1]/data_family['total']*100

In [34]:
data_family['percentage_of_non_return'] = data_family['percentage_of_non_return'].apply(lambda x: f"{x:.2f}%")

In [35]:
data_family['borrowers_share'] = (data_family['total']/ data_family['total'].sum())*100

In [36]:
data_family['borrowers_share'] = data_family['borrowers_share'].apply(lambda x: f"{x:.2f}%")

In [37]:
data_family.sort_values('percentage_of_non_return')

debt                       0    1  total percentage_of_non_return  \
family_status                                                       
вдовец / вдова           888   63    951                    6.62%   
в разводе               1105   84   1189                    7.06%   
женат / замужем        11334  927  12261                    7.56%   
гражданский брак        3749  385   4134                    9.31%   
Не женат / не замужем   2523  273   2796                    9.76%   

debt                  borrowers_share  
family_status                          
вдовец / вдова                  4.46%  
в разводе                       5.57%  
женат / замужем                57.48%  
гражданский брак               19.38%  
Не женат / не замужем          13.11%

**Вывод:** 
Анализ данных по семейному положению заемщиков показывает, что клиенты, состоящие в браке или ранее состоявшие в нём, имеют более низкий процент невозврата кредитов по сравнению с не женатыми/не замужними и клиентами в гражданском браке. Наиболее высокая доля невозврата наблюдается у клиентов, не состоящих в официальном браке. Это может быть связано с большей финансовой ответственностью и стабильностью у тех, кто состоит или состоял в браке.

#### 3.3 Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

In [38]:
data_income = data.pivot_table(index=['total_income_category'], columns='debt', values='total_income', aggfunc='count' )

Воспользуемся категоризацией доходов, которую сделали ранее где:
- 0–30000 — `'E'`;
- 30001–50000 — `'D'`;
- 50001–200000 — `'C'`;
- 200001–1000000 — `'B'`;
- 1000001 и выше — `'A'`.

In [39]:
data_income['total'] = data_income[0] + data_income[1]

In [40]:
data_income['percentage_of_non_return'] = data_income[1]/data_income['total']*100

In [41]:
data_income['percentage_of_non_return'] = data_income['percentage_of_non_return'].apply(lambda x: f"{x:.2f}%")

In [42]:
data_income['borrowers_share'] = (data_income['total']/ data_income['total'].sum())*100

In [43]:
data_income['borrowers_share'] = data_income['borrowers_share'].apply(lambda x: f"{x:.2f}%")

In [44]:
data_income.sort_values('percentage_of_non_return')

debt                       0     1  total percentage_of_non_return  \
total_income_category                                                
D                        328    21    349                    6.02%   
B                       4660   354   5014                    7.06%   
A                         23     2     25                    8.00%   
C                      14568  1353  15921                    8.50%   
E                         20     2     22                    9.09%   

debt                  borrowers_share  
total_income_category                  
D                               1.64%  
B                              23.51%  
A                               0.12%  
C                              74.64%  
E                               0.10%

In [45]:
data_income.loc[['B', 'C']]

debt                       0     1  total percentage_of_non_return  \
total_income_category                                                
B                       4660   354   5014                    7.06%   
C                      14568  1353  15921                    8.50%   

debt                  borrowers_share  
total_income_category                  
B                              23.51%  
C                              74.64%

**Вывод:** 
Категория C охватывает значительно больше заемщиков по сравнению с категорией B (15921 против 5014). Это может указывать на то, что в диапазоне доходов категории C находится большее количество потенциальных заемщиков. Также возможно, что кредиты более доступны для людей с доходом в диапазоне C, что приводит к большему количеству кредитов, выданных этой группе.

Процент невозвратов по кредитам в категории C составляет 8.50%, что выше по сравнению с категорией B, где этот показатель равен 7.06%. Это связано с большим количество выданных кредитов и соответсвенно большим количетсвом дополнительных факторов влияющих на оплату кредита.

Остальные категории в расчет не брал в связи с малым объемом данных.

Более высокий доход может способствовать лучшему управлению финансовыми обязательствами и более высокому уровню возвращаемости кредитов в срок. Это может быть связано с лучшими финансовыми ресурсами и планированием у клиентов с более высоким доходом.

#### 3.4 Как разные цели кредита влияют на его возврат в срок?

In [46]:
data_purpose = data.pivot_table(index=['purpose_category'], columns='debt', values='total_income', aggfunc='count' )

In [47]:
data_purpose['total'] = data_purpose[0] + data_purpose[1]

In [48]:
data_purpose['percentage_of_non_return'] = data_purpose[1]/data_purpose['total']*100

In [49]:
data_purpose['percentage_of_non_return'] = data_purpose['percentage_of_non_return'].apply(lambda x: f"{x:.2f}%")

In [50]:
data_purpose['borrowers_share'] = (data_purpose['total']/ data_purpose['total'].sum())*100

In [51]:
data_purpose['borrowers_share'] = data_purpose['borrowers_share'].apply(lambda x: f"{x:.2f}%")

In [52]:
data_purpose.sort_values('percentage_of_non_return')

debt                         0    1  total percentage_of_non_return  \
purpose_category                                                      
операции с недвижимостью  9971  780  10751                    7.26%   
проведение свадьбы        2130  183   2313                    7.91%   
получение образования     3619  369   3988                    9.25%   
операции с автомобилем    3879  400   4279                    9.35%   

debt                     borrowers_share  
purpose_category                          
операции с недвижимостью          50.40%  
проведение свадьбы                10.84%  
получение образования             18.70%  
операции с автомобилем            20.06%

**Вывод:** 
Цели кредитования влияют на вероятность возврата кредита в срок. Кредиты на операции с недвижимостью, хотя и составляют наибольшую долю, имеют самый низкий процент невозврата, что указывает на более осторожное отношение заемщиков. Напротив, кредиты на получение образования и операции с автомобилем показывают более высокий процент невозврата и могут потребовать дополнительных мер по управлению рисками.


#### 3.5 Приведите возможные причины появления пропусков в исходных данных.

*Ответ:* 
Пропуски в столбцах days_employed (общий трудовой стаж в днях) и total_income (ежемесячный доход) могут возникать по разным причинам:

Технический сбой при миграции данных: Проблемы при переносе данных из одной системы в другую могут привести к потере или неполному сохранению информации.

Некорректное внесение данных сотрудником/клиентом: Ошибки при вводе данных, такие как неправильный формат или пропуск полей, могут привести к отсутствию информации.

Непонимание клиента, как и куда вносить данные: Пользователи могут не понимать, какие данные нужно предоставить или как правильно их заполнить, что может привести к пропускам.

Несоответствие официального и реального дохода; страх перед налоговой: Некоторые пользователи могут не предоставлять полные данные о доходе из-за опасений перед налоговыми органами или несоответствия между официальным и фактическим доходом.

Принципиальное нежелание делиться конфиденциальной информацией: Некоторые пользователи могут сознательно избегать предоставления информации о доходе или трудовом стаже по соображениям конфиденциальности.


#### 3.6 Объясните, почему заполнить пропуски медианным значением — лучшее решение для количественных переменных.

*Ответ:* 

Медиана менее чувствительна к выбросам, чем среднее значение. Если в данных есть экстремальные значения (очень большие или очень маленькие), они могут искажать среднее значение.
Медиана помогает избежать ситуации, когда небольшое количество данных и выбросы могут значительно искажать результаты.
Это делает медиану предпочтительным выбором в большинстве случаев для количественных переменных.

### Шаг 4: общий вывод.

Данные
В рамках проекта была проведена работа с данными по задолженностям клиентов перед кредитной организацией. Данные включали следующие категории:

Количество детей у заемщиков
Семейное положение
Доходы заемщиков
Цели кредита
Выполненные шаги
Очистка данных

Удаление пропусков и аномалий.
Проверка данных на корректность и согласованность.
Анализ данных

Рассчет общего количества заемщиков в каждой категории.
Определение доли невозвращенных кредитов в каждой подгруппе.
Определение доли заемщиков в каждой категории от общего числа заемщиков.
Анализ данных
Анализ показывает, что количество детей, семейное положение, цель кредита и доходы заемщиков имеют значительное влияние на вероятность погашения кредита в срок.

Семейное положение: Клиенты со стабильным семейным положением чаще погашают кредиты вовремя, в то время как те, кто имеет нестабильные семейные отношения, сталкиваются с более высокими рисками непогашения кредитов.
Цель кредита: Кредиты на операции с недвижимостью показывают наименьший риск непогашения.
Количество детей: Наличие детей негативно отражается на способности клиентов оплачивать обязательства в срок, возможно, из-за дополнительных расходов на содержание иждивенцев.
Доходы заемщиков: Более высокий доход способствует лучшему управлению финансовыми обязательствами и более высокому уровню возвращаемости кредитов в срок.
Выводы
На основании анализа можно сделать вывод, что:

Идеальный клиент: Находится в официальном браке, бездетный, с доходом 200,001–1,000,000 руб., берущий кредит для операций с недвижимостью.
Нежелательный клиент: Холостой, имеющий двух и более детей, с доходом 50,001–200,000 руб., берущий кредит для операций с автомобилем.
Итоговые рекомендации
Улучшение сбора данных
Разработать и внедрить более детализированные формы для сбора данных о заемщиках, учитывая их семейное положение, количество детей, уровень дохода и цель кредита.
Обеспечить регулярное обновление и проверку актуальности данных.
Система скоринга
Включить в скоринг-модель переменные, показавшие высокую корреляцию с возвратом/невозвратом кредитов, такие как количество детей, семейное положение и цель кредита.
Использовать результаты анализа для формирования более точных профилей риска заемщиков.
Регулярно пересматривать и обновлять скоринг-модель на основе новых данных и аналитических выводов.
